### PR TITLE
Added basic dataset (SQuAD)

### PR DESCRIPTION
**Pull request by jeantux**
There are several things we should consider in this pull request (: #16

Should the method to download and load the file keep the name as "download" ?
```elixir
dataset = Scidata.Squad.download()
```
for example Hugging Face uses the following pattern:
  **dataset = load_dataset('squad')**

Maybe we could use something like:

```elixir
dataset = Scidata.Squad.load_dataset()
```

which brings us to the next point, what kind of return should this method have:

Tools like Pandas, Hugging Face..., use datasets, as in the example below:

```
Dataset({
     features: ['id', 'title', 'context', 'question', 'answers'],
     num_rows: 1057
})
```

In them it is possible to have various information such as existing columns, number of records and some facilitating methods.

In the changes that I sent, the return is a simple list that is loaded as soon as it is executed, given these considerations, I believe I didn't choose the best decisions.

But I would like to discuss this further with you.

### CODE REVIEW

File: lib/scidata/squad.ex [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                  ...
+                ]
+              }
+              ...
+            ]
+          }
+        ]
+      }
+  """
+  def info do
```
Comment (josevalim) [inline lib/scidata/squad.ex]: I think this is a good idea but we don't return this anywhere, so we should remain consistent, and not add this for now, and include a separate discussion for this. :)

**Comment by josevalim** [comment]
> Tools like Pandas, Hugging Face..., use datasets, as in the example below:

So far, we have been considering that we return a format that is easy enough to import into Nx or Explorer. For example, maybe we can do `Explorer.DataFrame.from_maps(Foo.download())`, and we have documented those with examples.

Maybe in the future we will add `load_df` and similar to multiple of those datasets, but I don't think we are quite there yet.

TL;DR - we currently use the docs to show how it integrates with other tools. :)

**Comment by jeantux** [comment]
I'm adding this dataset, but some columns are subsets.
I found the explorer project amazing and would like to keep the data compatible with it, you know a good way to overcome this barrier ?

below is an example of how this is handled in other tools:
image

File: lib/scidata/squad.ex [hunk 2]
```diff
@@ -0,0 +1,51 @@
+defmodule Scidata.Squad do
+  @moduledoc """
+  Module for downloading the [Squad dataset](https://rajpurkar.github.io/SQuAD-explorer).
```
Comment (t-rutten) [inline lib/scidata/squad.ex]:
```suggestion
  Module for downloading the [SQuAD1.1 dataset](https://rajpurkar.github.io/SQuAD-explorer).
```

File: lib/scidata/squad.ex [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  """
+
+  require Scidata.Utils
+  alias Scidata.Utils
+
+  @base_url "https://rajpurkar.github.io/SQuAD-explorer/dataset/"
+  @dataset_file "train-v1.1.json"
```
Comment (t-rutten) [inline lib/scidata/squad.ex]: We can also download the test file, `dev-v1.1.json`, and make it available with a `download_test/0` function :)

Comment (jeantux) [reply lib/scidata/squad.ex]: Good idea, i will do that!

**Comment by cigrainger** [comment]
> I'm adding this dataset, but some columns are subsets. I found the explorer project amazing and would like to keep the data compatible with it, you know a good way to overcome this barrier ?
> 
> below is an example of how this is handled in other tools: image

Unfortunately we don’t have a method for nested dataframes or similar in Explorer yet. It’s definitely in the radar, and I would like to take a similar approach to R and the tidyverse here, with nested dataframes: https://tidyr.tidyverse.org/articles/nest.html

I’ll need to see what polars can support on this front. But effectively it would be a series of dataframes.

**Comment by jeantux** [comment]
I'll try to generate a model that will be easier to import in the future to the explorer

**Comment by josevalim** [comment]
> I'll try to generate a model that will be easier to import in the future to the explorer

If the dataset is nested, then I would keep the dataset as is for now. We just need to wait a bit for the tooling to catch up, but we will get there. :)

**Comment by cigrainger** [comment]
The other option is that a 'nested dataframe' is really just two dataframes. It's just a denormalised table that can be normalised. The column that represents nesting can just be another table with an id column for joins.

File: lib/scidata/squad.ex [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+  require Scidata.Utils
+  alias Scidata.Utils
+
+  @base_url "https://rajpurkar.github.io/SQuAD-explorer/dataset/"
+  @train_dataset_file "train-v1.1.json"
+  @test_dataset_file "dev-v1.1.json"
+
+  @doc """
+  Downloads the SQuAD dataset.
+  ## Examples
```
Comment (josevalim) [inline lib/scidata/squad.ex:16]:
```suggestion

  ## Examples

```

File: lib/scidata/squad.ex [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+                           } ->
+      next_questions = qas |> Enum.map(& &1["question"])
+
+      next_answers =
+        qas |> Enum.map(&(&1["answers"] |> Utils.map_list_to_table(["answer_start", "text"])))
+
+      next_ids = qas |> Enum.map(& &1["id"])
+
+      next_contexts = Stream.repeatedly(fn -> context end) |> Enum.take(length(next_questions))
+      next_titles = Stream.repeatedly(fn -> title end) |> Enum.take(length(next_questions))
```
Comment (josevalim) [inline lib/scidata/squad.ex]:
```suggestion
      next_contexts = List.duplicate(context, length(next_questions))
      next_titles = List.duplicate(title, length(next_questions))
```

**Comment by jeantux** [comment]
Talking to @t-rutten, he suggested discussing the importance of keeping the `get_join_tables/1`.
If Explorer doesn't automatically detect the overlapping columns (To nested table).

Then he suggested that we could use something like this:
```elixir
{_, contexts, qas} = Squad.download() |> Squad.get_join_tables()

left = Explorer.DataFrame.from_map(contexts)
right = Explorer.DataFrame.from_map(qas)
Explorer.DataFrame.join(left, right)
```

File: lib/scidata/squad.ex [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+               answer: answers,
+               context_id: foreign_context_ids,
+               question: questions,
+               question_id: question_ids
+             } = curr_qa_acc
+           } ->
+          next_context_id = length(contexts) + 1
+
+          next_questions = qas |> Enum.map(& &1["question"])
+          next_answers = qas |> Enum.map(& &1["answers"])
```
Comment (t-rutten) [inline lib/scidata/squad.ex]: We need to extract parts of each answer that `Explorer.DataFrame` will accept (i.e. just the answer's start index and text) instead of the singleton list of maps in the `"answer"` key, so we can add `answer_text` and `answer_start` to the QA accumulator, `qa_acc`.

File: lib/scidata/squad.ex [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+              question: questions ++ next_questions,
+              question_id: question_ids ++ next_ids
+          }
+
+          {next_context_acc, next_qa_acc}
+        end
+      )
+
+    {next_titles, next_contexts, next_qas}
+  end
```
Comment (t-rutten) [inline lib/scidata/squad.ex]: Here's how we can revise the QA accumulator to add `answer_text` and `answer_start`:

```suggestion
  def get_join_tables(results) do
    {titles, contexts, qas} =
      results
      |> Enum.reduce(
        {
          %{title: [], title_id: []},
          %{context: [], context_id: [], title_id: []},
          %{
            context_id: [],
            question_id: [],
            question: [],
            answer_text: [],
            answer_start: []
          }
        },
        &to_tables/2
      )

    {reverse_and_flatten(titles), reverse_and_flatten(contexts), reverse_and_flatten(qas)}
  end

  defp to_tables(%{"paragraphs" => paragraphs, "title" => title}, acc) do
    {title_acc, context_acc, qa_acc} = acc

    %{title: titles, title_id: title_ids} = title_acc

    next_title_id = length(titles) + 1

    next_titles = %{
      title_acc
      | title: [title | titles],
        title_id: [next_title_id | title_ids]
    }

    {next_contexts, next_qas} =
      paragraphs
      |> Enum.reduce(
        {context_acc, qa_acc},
        fn %{"context" => context, "qas" => qas},
           {
             %{
               context: contexts,
               context_id: context_ids,
               title_id: title_ids
             } = curr_context_acc,
             %{
               context_id: foreign_context_ids,
               question: questions,
               question_id: question_ids,
               answer_start: answer_starts,
               answer_text: answer_texts,
             } = curr_qa_acc
           } ->
          next_context_id = length(contexts) + 1

          next_questions = qas |> Enum.map(& &1["question"])
          next_question_ids = qas |> Enum.map(& &1["id"])

          answers = qas |> Enum.map(& &1["answers"])
          # Each answer is a singleton
          next_answer_starts = answers |> Enum.map(&hd(&1)["answer_start"])
          next_answer_texts = answers |> Enum.map(&hd(&1)["text"])

          next_context_acc = %{
            curr_context_acc
            | context: [context | contexts],
              context_id: [next_context_id | context_ids],
              title_id: [next_title_id | title_ids]
          }

          next_foreign_context_ids =
            Stream.repeatedly(fn -> next_context_id end)
            |> Enum.take(length(next_questions))

          next_qa_acc = %{
            curr_qa_acc
            | context_id: [next_foreign_context_ids | foreign_context_ids],
              question: [next_questions | questions],
              question_id: [next_question_ids | question_ids],
              answer_start: [next_answer_starts | answer_starts],
              answer_text: [next_answer_texts | answer_texts]
          }

          {next_context_acc, next_qa_acc}
        end
      )

    {next_titles, next_contexts, next_qas}
  end

  defp reverse_and_flatten(acc) do
    Enum.reduce(acc, %{}, fn {k, v}, acc -> Map.put(acc, k, :lists.reverse(List.flatten(v))) end)
  end
```

**Comment by t-rutten** [comment]
@cigrainger, your suggestion to separate a "nested dataframe" into tables with join keys was a great one. I'm partial to keeping `Squad.get_join_tables` and removing `Squad.flatten` so that `DataFrame.join` handles all necessary row-duplication. What do you all think? cc: @jeantux @josevalim 

As for integrating scidata with explorer more explicitly, how should we proceed at this point? We could document the sketch Jean gave above (continued below), or we could add integration tests.

```elixir
entries = Scidata.Squad.download()
tables = Scidata.Squad.get_join_tables(entries)
{titles, contexts, qas} = tables

qa_df = Explorer.DataFrame.from_map(qas)

titles_df = Explorer.DataFrame.from_map(titles)
contexts_df = Explorer.DataFrame.from_map(contexts)
tc_df = Explorer.DataFrame.join(titles_df, contexts_df)

squad_df = Explorer.DataFrame.join(tc_df, qa_df)
```

**Comment by jeantux** [comment]
very good
I like the idea of keeping good documentation.

File: lib/scidata/squad.ex [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+                    "question" => "How many people protested on the San Francisco torch route?"
+                  },
+                  ...
+                ]
+              }
+              ...
+            ]
+          }
+        ]
+      }
```
Comment (t-rutten) [inline lib/scidata/squad.ex]:
```suggestion
      [
        %{
          "paragraphs" => [
            %{
              "context" => "Architecturally, the school has a...",
              "qas" => [
                %{
                  "answers" => [%{"answer_start" => 515, "text" => "Saint Bernadette Soubirous"}],
                  "id" => "5733be284776f41900661182",
                  "question" => "To whom did the..."
                }, ...
              ]
            }
          ],
          "title" => "University_of_Notre_Dame"
        }, ...
      ]
```

Let's preview the return list with the first example.

File: lib/scidata/squad.ex [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+  defp download_dataset(dataset_name) do
+    content =
+      Utils.get!(@base_url <> dataset_name).body
+      |> Jason.decode!()
+
+    content["data"]
+  end
+
+  def flatten(results) do
```
Comment (t-rutten) [inline lib/scidata/squad.ex]: Let's remove `flatten/1` and `process_example/2` and rename the remaining functions to use with Explorer.

File: lib/scidata/squad.ex [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        title: [next_titles | titles],
+        context: [next_contexts | contexts],
+        question: [next_questions | questions],
+        answer: [next_answers | answers]
+      }
+    end)
+    |> Enum.reduce(%{}, fn {k, v}, acc -> Map.put(acc, k, :lists.reverse(List.flatten(v))) end)
+  end
+
+  def get_join_tables(results) do
```
Comment (t-rutten) [inline lib/scidata/squad.ex]:
```suggestion
  def to_map(dataset) do
```

File: lib/scidata/squad.ex [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+            answer_start: []
+          }
+        },
+        &to_tables/2
+      )
+
+    {reverse_and_flatten(titles), reverse_and_flatten(contexts), reverse_and_flatten(qas)}
+  end
+
+  defp to_tables(%{"paragraphs" => paragraphs, "title" => title}, acc) do
```
Comment (t-rutten) [inline lib/scidata/squad.ex]:
```suggestion
  defp add_example(%{"paragraphs" => paragraphs, "title" => title}, acc) do
```

File: lib/scidata/squad.ex [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+  @doc """
+  Downloads the SQuAD dataset.
+
+  ## Examples
+
+      iex> Scidata.Squad.download()
+      %{
+        [
+          %{
+            "paragraphs" => [
```
Comment (t-rutten) [inline lib/scidata/squad.ex]: @josevalim do you think we should convert all maps in the train and test datasets here to atom maps? It would require a pass through the data but we could be more specific in typespecs for these functions and resulting maps would be slightly easier to work with.

Comment (josevalim) [reply lib/scidata/squad.ex]: I would keep them as strings I think? To reduce the overheard and because I think dataframes support it?

File: lib/scidata/squad.ex (hunk 9 above)
Comment (josevalim) [inline lib/scidata/squad.ex]: I think some of those need to be private or removed? :)

Comment (jeantux) [reply lib/scidata/squad.ex]: some of these functions?
did you mean `process_example` and `to_map` ?

File: lib/scidata/squad.ex [hunk 13]
```diff
@@ -54,46 +78,32 @@ defmodule Scidata.Squad do
...
-      next_titles = List.duplicate(title, length(next_questions))
-
-      %{
-        id: [next_ids | ids],
-        title: [next_titles | titles],
-        context: [next_contexts | contexts],
-        question: [next_questions | questions],
-        answer: [next_answers | answers]
+  ## Examples
+    iex> Scidata.Squad.to_maps(entries)
```
Comment (josevalim) [inline lib/scidata/squad.ex]:
```suggestion

      iex> Scidata.Squad.to_maps(entries)
```

File: lib/scidata/squad.ex [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+          next_context_acc = %{
+            curr_context_acc
+            | context: [context | contexts],
+              context_id: [next_context_id | context_ids],
+              title_id: [next_title_id | title_ids]
+          }
+
+          next_foreign_context_ids =
+            Stream.repeatedly(fn -> next_context_id end)
+            |> Enum.take(length(next_questions))
```
Comment (josevalim) [inline lib/scidata/squad.ex]:
```suggestion
            List.duplicate(next_context_id, length(next_questions))
```

File: lib/scidata/squad.ex [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+  defp download_dataset(dataset_name) do
+    content =
+      Utils.get!(@base_url <> dataset_name).body
+      |> Jason.decode!()
+
+    content["data"]
+  end
+
+  @doc """
+  Converts the squad dataset to a tuple containing three maps.
```
Comment (josevalim) [inline lib/scidata/squad.ex]: Can you please expand what is on those there maps?

**Comment by josevalim** [comment]
Hi folks, sorry for the back and forth in this, but this is the first dataset of is type!

After looking at the code (and the example, which is a great addition), I am now thinking that going with `flatten` may be the best approach. The reason why I think so is because we are already somewhat kind of denormalizing by adding the IDs? Plus using flatten means the code is simpler both internally and in the data-frames usage. The code for flatten would be:

```elixir
for %{"paragraph" => paragraph, ...all_fields_at_the_root...} <- results,
    %{"context" => context, "qas" => qas} <- paragraph,
    qa <- qas do
  Map.merge(qa, %{"context" => context, "paragraph" => paragraph, ...all_fields_at_the_root...})
end
```

And that should be almost it. Thoughts?

**Comment by jeantux** [comment]
I think it's a good idea to have the flatten function, I've added it, but if you change your mind, no problem.
If we decide to keep it, I can create some tests and document them with examples. (:

For the cases of nested tables, leaving everything flatten seems to solve.

File: lib/scidata/squad.ex [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    normalized_data
+    |> Enum.reduce(table, fn map, acc ->
+      Enum.map(acc, fn {key, values} ->
+        {key, [map[key] | values]}
+      end)
+    end)
+    |> Enum.map(fn {key, values} -> {key, :lists.reverse(values)} end)
+    |> Enum.into(%{})
```
Comment (josevalim) [inline lib/scidata/squad.ex]: I think we might not need this. What if we DF.from_list (or similar) instead of DF.from_map? They should have an API that accepts lists of rows?

Comment (t-rutten) [reply lib/scidata/squad.ex]: I see just `from_map/1` on [the Explorer docs](https://elixir-nx.github.io/explorer/Explorer.DataFrame.html#from_map/2) right now.

We can make the whole map inside the `for` loop though: https://gist.github.com/t-rutten/17cd76074ecdd3d6062c631a8a749963

Comment (josevalim) [reply lib/scidata/squad.ex]: @t-rutten nice!

So final suggestion: let's go with @t-rutten code above, call it `to_map` (instead of `flatten`), remove the existing `to_maps`, and ship it?

I am so sorry for all the back and forth in this. :D

Comment (josevalim) [reply lib/scidata/squad.ex]: Oh, what if we call it `to_columns` instead of `to_map`? It will make it clear we are returning the columns.

Comment (t-rutten) [reply lib/scidata/squad.ex]: Perfect :)

Comment (t-rutten) [reply lib/scidata/squad.ex]: I kind of like `to_map` so it's `Scidata.to_map` -> `Explorer.from_map`, but `to_columns` totally works. Do you have a preference @jeantux?

Comment (jeantux) [reply lib/scidata/squad.ex]: `to_columns` sounds good to me

**Review by josevalim — APPROVED** [review]

File: lib/scidata/squad.ex [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+      %{
+        "answer_start" => [515, ...],
+        "context" => ["Architecturally, the...", ...],
+        "id" => ["5733be284776f41900661182", ...],
+        "question" => ["To whom did the Vir...", ...],
+        "answer_text" => ["Saint Bernadette Soubirous", ...],
+        "title" => ["University_of_Notre_Dame", ...]
+      }
+
+  """
```
Comment (t-rutten) [inline lib/scidata/squad.ex]:
```suggestion
  @doc """
  Convert result of `download/0` or `download_test/0` to map for use with `Explorer.DataFrame`.

  ## Examples

      iex> columns_for_df = Scidata.Squad.download() |> Scidata.Squad.to_columns()
      %{
        "answer_start" => [515, ...],
        "context" => ["Architecturally, the...", ...],
        "id" => ["5733be284776f41900661182", ...],
        "question" => ["To whom did the Vir...", ...],
        "answer_text" => ["Saint Bernadette Soubirous", ...],
        "title" => ["University_of_Notre_Dame", ...]
      }
      iex> Explorer.DataFrame.from_map(columns_for_df)
      #Explorer.DataFrame<
      [rows: 87599, columns: 6]
      ...
      >
  """
```
Last suggestion just for documentation. Because `to_columns` is much simpler now, we can remove the example file.

**Review by t-rutten — APPROVED** [review]
